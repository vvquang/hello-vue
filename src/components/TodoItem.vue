<template>
  <div class="todo-item">
    <div class="todo-item-content">
      <input
        type="checkbox"
        v-model="completed"
        class="todo-item-checkbox"
        @change="handleDoneEdit"
      />
      <div>{{ index + 1 }}.</div>
      <div
        v-if="!editing"
        @dblclick="handleEditTodo"
        class="todo-item-label"
        :class="{ completed: completed }"
      >
        {{ title }}
      </div>
      <input
        v-else
        class="todo-item-edit"
        type="text"
        v-model="title"
        @blur="handleDoneEdit"
        @keyup.enter="handleDoneEdit"
        @keyup.esc="handleCancelEdit"
        v-focus
      />
    </div>
    <div class="remove-item" @click="handleRemoveTodo(index)">
      &times;
    </div>
  </div>
</template>

<script>
export default {
  name: "todo-item",

  props: {
    todo: {
      type: Object,
      required: true,
    },
    index: {
      type: Number,
      required: true,
    },
    checkAll: {
      type: Boolean,
      required: true,
    },
  },

  data() {
    return {
      id: this.todo.id,
      title: this.todo.title,
      completed: this.todo.completed,
      editing: this.todo.editing,
      beforeEditCache: "",
    };
  },

  watch: {
    checkAll() {
      this.completed = this.checkAll ? true : this.todo.completed;
    },
  },

  directives: {
    focus: {
      inserted: function(el) {
        el.focus();
      },
    },
  },

  methods: {
    handleRemoveTodo(index) {
      this.$emit("removedTodo", index);
    },

    handleEditTodo() {
      this.beforeEditCache = this.title;
      this.editing = true;
    },

    handleDoneEdit() {
      if (this.title.trim() == "") {
        this.title = this.beforeEditCache;
      }
      this.editing = false;

      this.$emit("finishedEdit", {
        index: this.index,
        todo: {
          id: this.id,
          title: this.title,
          completed: this.completed,
          editing: this.editing,
        },
      });
    },

    handleCancelEdit() {
      this.title = this.beforeEditCache;
      this.editing = false;
    },
  },
};
</script>

<style scoped></style>
