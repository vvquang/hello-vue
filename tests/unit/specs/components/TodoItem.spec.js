import { shallowMount } from '@vue/test-utils'

import TodoItem from '@/components/TodoItem'
import wrapperOps from '../../supports/wrapper'
import { todo } from '../../supports/fixtures/todo-item'

const REMVOVE_TODO = jest.spyOn(TodoItem.methods, 'REMVOVE_TODO')
const CHECK_ITEM_TODO = jest.spyOn(TodoItem.methods, 'CHECK_ITEM_TODO')
const FINISH_EDIT_TODO = jest.spyOn(TodoItem.methods, 'FINISH_EDIT_TODO')

const handleRemoveTodo = jest.spyOn(TodoItem.methods, 'handleRemoveTodo')
const handleEditTodo = jest.spyOn(TodoItem.methods, 'handleEditTodo')
const handleCompleteItem = jest.spyOn(TodoItem.methods, 'handleCompleteItem')
const handleDoneEdit = jest.spyOn(TodoItem.methods, 'handleDoneEdit')
const handleCancelEdit = jest.spyOn(TodoItem.methods, 'handleCancelEdit')

describe('TodoItem.vue', () => {
  let wrapper

  afterEach(() => {
    wrapper.vm.$destroy()
  })

  const localWrapperOps = {
    ...wrapperOps,
    propsData: {
      todo,
      index: 0,
      checkAll: false
    },
    computed: {}
  }

  beforeEach(() => {
    wrapper = shallowMount(TodoItem, localWrapperOps)
  })

  it('component should render correctly', () => {
    expect(wrapper.exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should call handleRemoveTodo', async () => {
    wrapper.find('.remove-item').trigger('click')
    await wrapper.vm.$nextTick()

    expect(handleRemoveTodo).toHaveBeenCalled()
    expect(REMVOVE_TODO).toHaveBeenCalledWith(0)
  })

  it('should call handleEditTodo', async () => {
    wrapper.find('.todo-item-label').trigger('dblclick')
    await wrapper.vm.$nextTick()
    expect(handleEditTodo).toHaveBeenCalled()
  })

  it('should call handleDoneEdit', async () => {
    // enable input
    wrapper.find('.todo-item-label').trigger('dblclick')
    await wrapper.vm.$nextTick()

    wrapper.find('.todo-item-edit').trigger('keyup.enter')
    await wrapper.vm.$nextTick()

    expect(handleDoneEdit).toHaveBeenCalled()
    expect(FINISH_EDIT_TODO).toHaveBeenCalledWith({
      index: 0,
      todo: todo
    })
  })

  it('should call handleCompleteItem', async () => {
    const cbxComplete = wrapper.find('input[type="checkbox"]')
    await cbxComplete.trigger('click')

    expect(handleCompleteItem).toHaveBeenCalled()
    expect(CHECK_ITEM_TODO).toHaveBeenCalledWith({
      id: todo.id,
      completed: true
    })
  })

  it('should call handleCancelEdit', async () => {
    // enable input
    wrapper.find('.todo-item-label').trigger('dblclick')
    await wrapper.vm.$nextTick()

    wrapper.find('.todo-item-edit').trigger('keyup.esc')
    await wrapper.vm.$nextTick()

    expect(handleCancelEdit).toHaveBeenCalled()
  })
})
