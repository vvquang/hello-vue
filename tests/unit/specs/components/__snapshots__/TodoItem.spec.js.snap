// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem.vue component should render correctly 1`] = `
<div class="todo-item">
  <div class="todo-item-content"><input type="checkbox" class="todo-item-checkbox">
    <div>1.</div>
    <div class="todo-item-label">
      Finish vue
    </div>
  </div>
  <div class="remove-item">
    Ã—
  </div>
</div>
`;
